warehouse/query-core/src/main/java/datawave/query/util/sortedmap/BufferedFileBackedSortedMap.java

```diff
@@ -0,0 +1,560 @@
+    private static final Logger log = Logger.getLogger(BufferedFileBackedSortedMap.class);
+    protected static final int DEFAULT_BUFFER_PERSIST_THRESHOLD = 1000;
+    protected static final int DEFAULT_MAX_OPEN_FILES = 100;
+    protected static final int DEFAULT_NUM_RETRIES = 2;
+
+    protected MultiMapBackedSortedMap<K,V> map = new MultiMapBackedSortedMap<>();
+    protected int maxOpenFiles = DEFAULT_MAX_OPEN_FILES;
+    protected FileSortedMap<K,V> buffer = null;
+    protected FileSortedMap.FileSortedMapFactory<K,V> mapFactory = null;
+    protected final Comparator<K> comparator;
+    protected FileSortedMap.RewriteStrategy<K,V> rewriteStrategy;
+    protected boolean sizeModified = false;
+    protected int size = 0;
+    protected int numRetries = DEFAULT_NUM_RETRIES;
+
+    protected List<SortedMapFileHandlerFactory> handlerFactories;
+    protected int bufferPersistThreshold = DEFAULT_BUFFER_PERSIST_THRESHOLD;
+
+    /**
+     * A factory for SortedMapFileHandlers
+     *
+     *
+     *
+     */
+    public interface SortedMapFileHandlerFactory {
+        SortedMapFileHandler createHandler() throws IOException;
+
+        boolean isValid();
+    }
+
+    public static class Builder<B extends Builder<B,K,V>,K,V> {
+        private int maxOpenFiles = DEFAULT_MAX_OPEN_FILES;
+        private FileSortedMap.FileSortedMapFactory<K,V> mapFactory = new FileSerializableSortedMap.Factory();
+        private Comparator<K> comparator;
+        private FileSortedMap.RewriteStrategy<K,V> rewriteStrategy;
+        private int numRetries = DEFAULT_NUM_RETRIES;
+        private List<SortedMapFileHandlerFactory> handlerFactories = new ArrayList<>();
+        private int bufferPersistThreshold = DEFAULT_BUFFER_PERSIST_THRESHOLD;
+
+        public Builder() {}
+
+        @SuppressWarnings("unchecked")
+        protected B self() {
+            return (B) this;
+        }
+
+        public B withMaxOpenFiles(int maxOpenFiles) {
+            this.maxOpenFiles = maxOpenFiles;
+            return self();
+        }
+
+        @SuppressWarnings("unchecked")
+        public B withMapFactory(FileSortedMap.FileSortedMapFactory<?,?> mapFactory) {
+            this.mapFactory = (FileSortedMap.FileSortedMapFactory<K,V>) mapFactory;
+            return self();
+        }
+
+        @SuppressWarnings("unchecked")
+        public B withComparator(Comparator<?> comparator) {
+            this.comparator = (Comparator<K>) comparator;
+            return self();
+        }
+
+        @SuppressWarnings("unchecked")
+        public B withRewriteStrategy(FileSortedMap.RewriteStrategy<?,?> rewriteStrategy) {
+            this.rewriteStrategy = (FileSortedMap.RewriteStrategy<K,V>) rewriteStrategy;
+            return self();
+        }
+
+        public B withNumRetries(int numRetries) {
+            this.numRetries = numRetries;
+            return self();
+        }
+
+        public B withHandlerFactories(List<SortedMapFileHandlerFactory> handlerFactories) {
```

Comment:
this method is never called. Not sure this class will work right without the handlerFactories being set

Reply:
yikes... that does not sound good.  I will figure that out.

Reply:
Ah, I see, the handler factories are setup in the HdfsBackedSortedMap which extends the BufferedFileBackedSortedMap.  This is the builder that the UniqueTransform uses.

Reply:
my mistake, sorry